README
IM Status - Rhythmbox plugin - 0.7
==================================

Author: Vincent Untz <vuntz@gnome.org>
Licence: GPLv2
Main bzr branch: http://www.vuntz.net/bzr/rhythmbox-plugins/

This Rhythmbox plugin updates the status in various IM clients (Empathy and
Gossip are currently supported) according to the current playing song.

Here is the list of known bugs:

 + with Gossip <= 0.22: when you deactivate the plugin, or quit Rhythmbox,
   your status is set to "". This has been fixed in Gossip. See
   http://bugzilla.gnome.org/show_bug.cgi?id=410398 for details.

 + with Empathy and some old version of mission-control: the plugin doesn't
   work because mission-control didn't emit a signal when the status message is
   changed but the status itself doesn't change. This has been fixed in later
   releases of mission-control.

** Installation **

To install this plugin, you only have to copy the plugin directory to
~/.gnome2/rhythmbox/plugins/im-status

It should just work when you enable it in the plugin dialog of Rhythmbox.

** Reporting bugs **

Please report bugs by sending a mail to the author(s).

** Hacking **

Feel free to create a bzr branch or send patch to the author(s).


NEWS
0.7
===

Add support for Empathy.
Rename to im-status.

0.6
===

Resave the gossip status when returning from pause, because the user might
have changed it since he activated the plugin.

0.5
===

Correctly update the status when returning from pause.

0.4
===

Update README to explain the known "bug" caused by Gossip not correctly
dealing with an empty status.
When stopping playing, reset the status (since the user is no longer listening)
Save the original Gossip status, and use it when reseting the status.
Don't uselessly change the status when a song property is changed, but
it wouldn't affect the status.

0.3
===

Make sure the plugin works with latest dbus-python bindings.
Rename init.py to __init__.py so that people can use the plugin :-)

0.2
===

Update README.
Rename gossip.py to init.py.

0.1
===

Initial release


init.py
# coding: utf-8
# vim: set et sw=2:
# 
# Copyright (C) 2007 - Vincent Untz
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2, or (at your option)
# any later version.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301  USA.

import rhythmdb, rb
import dbus
from dbus import dbus_bindings

BUS_NAME = 'org.gnome.Gossip'
OBJ_PATH = '/org/gnome/Gossip'
IFACE_NAME = 'org.gnome.Gossip'

class GossipStatusPlugin (rb.Plugin):
  def __init__ (self):
    rb.Plugin.__init__ (self)

  def activate (self, shell):
    self.shell = shell
    sp = shell.get_player ()
    self.psc_id  = sp.connect ('playing-song-changed',
                               self.playing_entry_changed)
    self.pc_id   = sp.connect ('playing-changed',
                               self.playing_changed)
    self.pspc_id = sp.connect ('playing-song-property-changed',
                               self.playing_song_property_changed)

    self.current_entry = None

    entry = sp.get_playing_entry ()
    self.playing_entry_changed (sp, entry)

  def deactivate (self, shell):
    self.shell = None
    sp = shell.get_player ()
    sp.disconnect (self.psc_id)
    sp.disconnect (self.pc_id)
    sp.disconnect (self.pspc_id)

    if self.current_entry is not None:
      self.set_status ("")

  def playing_changed (self, sp, playing):
    self.set_entry (sp.get_playing_entry ())

  def playing_entry_changed (self, sp, entry):
    self.set_entry (entry)

  def playing_song_property_changed (self, sp, uri, property, old, new):
    self.set_status_from_entry ()

  def set_entry (self, entry):
    if entry == self.current_entry:
      return

    self.current_entry = entry

    if entry is None:
      self.set_status ("")
      return

    self.set_status_from_entry ()

  def set_status_from_entry (self):
    db = self.shell.get_property ("db")
    artist = db.entry_get (self.current_entry, rhythmdb.PROP_ARTIST) or _("Unknown")
    song   = db.entry_get (self.current_entry, rhythmdb.PROP_TITLE) or _("Unknown")

    new_status = _(u"♫ %s - %s ♫") % (artist, song)
    self.set_status (new_status)

  def set_status (self, new_status):
    try:
      bus = dbus.SessionBus ()
      gossip_obj = bus.get_object (BUS_NAME, OBJ_PATH)
      gossip = dbus.Interface (gossip_obj, IFACE_NAME)

      state, status = gossip.GetPresence ("")
      gossip.SetPresence (state, new_status)
    except dbus_bindings.DBusException:
      pass


__init__.py
# coding: utf-8
# vim: set et sw=2:
# 
# Copyright (C) 2007-2008 - Vincent Untz
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2, or (at your option)
# any later version.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301  USA.

import rhythmdb, rb
try:
  import dbus
  use_gossip = True
except ImportError:
  use_gossip = False
try:
  import empathy
  use_empathy = True
  empathy_idle = empathy.Idle ()
except ImportError:
  use_empathy = False

BUS_NAME = 'org.gnome.Gossip'
OBJ_PATH = '/org/gnome/Gossip'
IFACE_NAME = 'org.gnome.Gossip'

class IMStatusPlugin (rb.Plugin):
  def __init__ (self):
    rb.Plugin.__init__ (self)

  def activate (self, shell):
    self.shell = shell
    sp = shell.get_player ()
    self.psc_id  = sp.connect ('playing-song-changed',
                               self.playing_entry_changed)
    self.pc_id   = sp.connect ('playing-changed',
                               self.playing_changed)
    self.pspc_id = sp.connect ('playing-song-property-changed',
                               self.playing_song_property_changed)

    self.current_entry = None

    self.save_status ()

    if sp.get_playing ():
      self.set_entry (sp.get_playing_entry ())

  def deactivate (self, shell):
    self.shell = None
    sp = shell.get_player ()
    sp.disconnect (self.psc_id)
    sp.disconnect (self.pc_id)
    sp.disconnect (self.pspc_id)

    if self.current_entry is not None:
      self.restore_status ()

  def playing_changed (self, sp, playing):
    if playing:
      self.set_entry (sp.get_playing_entry ())
    else:
      self.current_entry = None
      self.restore_status ()

  def playing_entry_changed (self, sp, entry):
    if sp.get_playing ():
      self.set_entry (entry)

  def playing_song_property_changed (self, sp, uri, property, old, new):
    if sp.get_playing () and (property == "artist" or property == "title"):
      self.set_status_from_entry ()

  def set_entry (self, entry):
    if entry == self.current_entry:
      return

    if self.current_entry == None:
      self.save_status ()
    self.current_entry = entry

    if entry is None:
      self.restore_status ()
      return

    self.set_status_from_entry ()

  def set_status_from_entry (self):
    db = self.shell.get_property ("db")
    artist = db.entry_get (self.current_entry, rhythmdb.PROP_ARTIST) or _("Unknown")
    song   = db.entry_get (self.current_entry, rhythmdb.PROP_TITLE) or _("Unknown")

    new_status = _(u"♫ %s - %s ♫") % (artist, song)
    self.set_status (new_status)

  def set_status (self, new_status):
    self.set_gossip_status (new_status)
    self.set_empathy_status (new_status)

  def save_status (self):
    self.saved_gossip = self.get_gossip_status ()
    self.saved_empathy = self.get_empathy_status ()

  def restore_status (self):
    if self.saved_gossip != None:
      self.set_gossip_status (self.saved_gossip)
    if self.saved_empathy != None:
      self.set_empathy_status (self.saved_empathy)

  def set_gossip_status (self, new_status):
    if not use_gossip:
      return

    try:
      bus = dbus.SessionBus ()
      gossip_obj = bus.get_object (BUS_NAME, OBJ_PATH)
      gossip = dbus.Interface (gossip_obj, IFACE_NAME)

      state, status = gossip.GetPresence ("")
      gossip.SetPresence (state, new_status)
    except dbus.DBusException:
      pass

  def get_gossip_status (self):
    if not use_gossip:
      return

    try:
      bus = dbus.SessionBus ()
      gossip_obj = bus.get_object (BUS_NAME, OBJ_PATH)
      gossip = dbus.Interface (gossip_obj, IFACE_NAME)

      state, status = gossip.GetPresence ("")
      return status
    except dbus.DBusException:
      return None

  def set_empathy_status (self, new_status):
    if not use_empathy:
      return

    empathy_idle.set_status (new_status)

  def get_empathy_status (self):
    if not use_empathy:
      return

    return empathy_idle.get_status ()
